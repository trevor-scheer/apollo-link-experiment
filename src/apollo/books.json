{"data":{"allBooks":[{"title":"Functional JavaScript: Introducing Functional Programming with Underscore.js","__typename":"Book"},{"title":"Maintainable JavaScript - Writing Readable Code","__typename":"Book"},{"title":"Professional JavaScript for Web Developers","__typename":"Book"},{"title":"JavaScript Patterns","__typename":"Book"},{"title":"JavaScript: The Definitive Guide","__typename":"Book"},{"title":"Secrets of the JavaScript Ninja","__typename":"Book"},{"title":"Object-Oriented Javascript","__typename":"Book"},{"title":"JavaScript: The Good Parts","__typename":"Book"},{"title":"JavaScript & jQuery: The Missing Manual","__typename":"Book"},{"title":"High Performance JavaScript","__typename":"Book"},{"title":"Eloquent JavaScript: A Modern Introduction to Programming","__typename":"Book"},{"title":"Pro JavaScript Techniques","__typename":"Book"},{"title":"High Performance Web Sites: Essential Knowledge for Front-End Engineers","__typename":"Book"},{"title":"Even Faster Web Sites","__typename":"Book"},{"title":"Learning JavaScript Design Patterns","__typename":"Book"},{"title":"jQuery in Action","__typename":"Book"},{"title":"jQuery Pocket Reference","__typename":"Book"},{"title":"Book of Speed [Online]","__typename":"Book"},{"title":"JavaScript Allonge [Ebook]","__typename":"Book"},{"title":"JavaScript for PHP Developers","__typename":"Book"},{"title":"Effective JavaScript: 68 Specific Ways to Harness the Power of JavaScript","__typename":"Book"}]}}