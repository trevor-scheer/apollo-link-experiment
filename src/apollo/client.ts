import ApolloClient from 'apollo-client';

import {InMemoryCache} from 'apollo-cache-inmemory';
import {createHttpLink} from 'apollo-link-http';
import {ApolloLink, from, Observable} from 'apollo-link';
import {fstat, existsSync, writeFileSync} from 'fs';
import {resolve} from 'path';

const cache = new InMemoryCache();

const readWriteLink = new ApolloLink((operation, forward) => {
  const fileName = resolve(__dirname, `${operation.operationName}.json`);

  if (existsSync(fileName)) {
    const data = require(fileName);
    return new Observable(observer => {
      console.log('returning JSON');
      observer.next(data);
      observer.complete();
    });
  }
  console.log('returning network request, writing file');

  return forward!(operation).map(data => {
    if (!existsSync(fileName)) {
      writeFileSync(
        resolve(__dirname, `${operation.operationName}.json`),
        JSON.stringify(data)
      );
    }

    return data;
  });
});

const httpLink = createHttpLink({
  uri: 'https://graphql-demo-v2.now.sh/'
});

const link = from([readWriteLink, httpLink]);

export const client = new ApolloClient({
  link,
  cache
});
