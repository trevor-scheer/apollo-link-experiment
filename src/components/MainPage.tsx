import React, {Component} from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';

const people = gql`
  query books {
    allBooks {
      title
    }
  }
`;

export class MainPage extends Component {
  render() {
    return (
      <Query query={people}>
        {({data, loading, error}) => {
          if (loading) return null;
          if (error) {
            console.log(error);
            return null;
          }

          // console.log({data});
          return <div>Hello World</div>;
        }}
      </Query>
    );
  }
}
