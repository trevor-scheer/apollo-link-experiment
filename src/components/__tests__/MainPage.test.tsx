import React from 'react';
import {MainPage} from '../MainPage';
import {ApolloProvider, getDataFromTree} from 'react-apollo';
import {client} from '../../apollo/client';

it('renders without crashing', async () => {
  await getDataFromTree(
    <ApolloProvider client={client}>
      <MainPage />
    </ApolloProvider>
  );

  expect(client.extract()).toMatchSnapshot();
  return;
});
