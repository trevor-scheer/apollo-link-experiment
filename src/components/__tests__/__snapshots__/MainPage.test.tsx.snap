// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Object {
  "ROOT_QUERY": Object {
    "allBooks": Array [
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.0",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.1",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.2",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.3",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.4",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.5",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.6",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.7",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.8",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.9",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.10",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.11",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.12",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.13",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.14",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.15",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.16",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.17",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.18",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.19",
        "type": "id",
        "typename": "Book",
      },
      Object {
        "generated": true,
        "id": "ROOT_QUERY.allBooks.20",
        "type": "id",
        "typename": "Book",
      },
    ],
  },
  "ROOT_QUERY.allBooks.0": Object {
    "__typename": "Book",
    "title": "Functional JavaScript: Introducing Functional Programming with Underscore.js",
  },
  "ROOT_QUERY.allBooks.1": Object {
    "__typename": "Book",
    "title": "Maintainable JavaScript - Writing Readable Code",
  },
  "ROOT_QUERY.allBooks.10": Object {
    "__typename": "Book",
    "title": "Eloquent JavaScript: A Modern Introduction to Programming",
  },
  "ROOT_QUERY.allBooks.11": Object {
    "__typename": "Book",
    "title": "Pro JavaScript Techniques",
  },
  "ROOT_QUERY.allBooks.12": Object {
    "__typename": "Book",
    "title": "High Performance Web Sites: Essential Knowledge for Front-End Engineers",
  },
  "ROOT_QUERY.allBooks.13": Object {
    "__typename": "Book",
    "title": "Even Faster Web Sites",
  },
  "ROOT_QUERY.allBooks.14": Object {
    "__typename": "Book",
    "title": "Learning JavaScript Design Patterns",
  },
  "ROOT_QUERY.allBooks.15": Object {
    "__typename": "Book",
    "title": "jQuery in Action",
  },
  "ROOT_QUERY.allBooks.16": Object {
    "__typename": "Book",
    "title": "jQuery Pocket Reference",
  },
  "ROOT_QUERY.allBooks.17": Object {
    "__typename": "Book",
    "title": "Book of Speed [Online]",
  },
  "ROOT_QUERY.allBooks.18": Object {
    "__typename": "Book",
    "title": "JavaScript Allonge [Ebook]",
  },
  "ROOT_QUERY.allBooks.19": Object {
    "__typename": "Book",
    "title": "JavaScript for PHP Developers",
  },
  "ROOT_QUERY.allBooks.2": Object {
    "__typename": "Book",
    "title": "Professional JavaScript for Web Developers",
  },
  "ROOT_QUERY.allBooks.20": Object {
    "__typename": "Book",
    "title": "Effective JavaScript: 68 Specific Ways to Harness the Power of JavaScript",
  },
  "ROOT_QUERY.allBooks.3": Object {
    "__typename": "Book",
    "title": "JavaScript Patterns",
  },
  "ROOT_QUERY.allBooks.4": Object {
    "__typename": "Book",
    "title": "JavaScript: The Definitive Guide",
  },
  "ROOT_QUERY.allBooks.5": Object {
    "__typename": "Book",
    "title": "Secrets of the JavaScript Ninja",
  },
  "ROOT_QUERY.allBooks.6": Object {
    "__typename": "Book",
    "title": "Object-Oriented Javascript",
  },
  "ROOT_QUERY.allBooks.7": Object {
    "__typename": "Book",
    "title": "JavaScript: The Good Parts",
  },
  "ROOT_QUERY.allBooks.8": Object {
    "__typename": "Book",
    "title": "JavaScript & jQuery: The Missing Manual",
  },
  "ROOT_QUERY.allBooks.9": Object {
    "__typename": "Book",
    "title": "High Performance JavaScript",
  },
}
`;
